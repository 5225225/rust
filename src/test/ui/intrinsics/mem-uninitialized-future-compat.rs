#![feature(never_type)]
#![deny(mem_uninitialized)]
#![allow(deprecated, invalid_value, dead_code)]

use std::mem::MaybeUninit;

struct UninitStruct {
    a: u32,
    b: char,
}

enum OneVariant {
    Hello,
}

enum TwoVariant {
    Hello,
    Goodbye,
}

enum OneVariantWith<T> {
    Hello(T),
}

unsafe fn unknown_type<T, const N: usize>() {
    std::mem::uninitialized::<T>();
    //~^ ERROR the type `T` is generic, and might not permit being left uninitialized

    std::mem::uninitialized::<[T; N]>();
    //~^ ERROR the type `[T; N]` is generic, and might not permit being left uninitialized

    std::mem::uninitialized::<[char; N]>();
    //~^ ERROR the type `[char; N]` is generic, and might not permit being left uninitialized

    std::mem::uninitialized::<[UninitStruct; N]>();
    //~^ ERROR the type `[UninitStruct; N]` is generic, and might not permit being left uninitialized

    std::mem::uninitialized::<Result<T, !>>();
    //~^ ERROR the type `std::result::Result<T, !>` does not permit being left uninitialized

    std::mem::uninitialized::<OneVariantWith<T>>();
    //~^ ERROR the type `OneVariantWith<T>` is generic, and might not permit being left uninitialized

    std::mem::uninitialized::<[T; 0]>();
    std::mem::uninitialized::<[char; 0]>();
}

fn main() {
    unsafe {
        std::mem::uninitialized::<&'static u32>();
        //~^ ERROR the type `&u32` does not permit being left uninitialized

        std::mem::uninitialized::<Box<u32>>();
        //~^ ERROR the type `std::boxed::Box<u32>` does not permit being left uninitialized

        std::mem::uninitialized::<fn()>();
        //~^ ERROR the type `fn()` does not permit being left uninitialized

        std::mem::uninitialized::<!>();
        //~^ ERROR the type `!` does not permit being left uninitialized

        std::mem::uninitialized::<*mut dyn std::io::Write>();
        //~^ ERROR the type `*mut dyn std::io::Write` does not permit being left uninitialized

        std::mem::uninitialized::<bool>();
        //~^ ERROR the type `bool` does not permit being left uninitialized

        std::mem::uninitialized::<char>();
        //~^ ERROR the type `char` does not permit being left uninitialized

        std::mem::uninitialized::<UninitStruct>();
        //~^ ERROR the type `UninitStruct` does not permit being left uninitialized

        std::mem::uninitialized::<[UninitStruct; 16]>();
        //~^ ERROR the type `[UninitStruct; 16]` does not permit being left uninitialized

        std::mem::uninitialized::<(u32, char)>();
        //~^ ERROR the type `(u32, char)` does not permit being left uninitialized

        std::mem::uninitialized::<TwoVariant>();
        //~^ ERROR the type `TwoVariant` does not permit being left uninitialized

        std::mem::uninitialized::<Result<!, !>>();
        //~^ ERROR the type `std::result::Result<!, !>` does not permit being left uninitialized

        std::mem::uninitialized::<Result<!, u32>>();
        //~^ ERROR the type `std::result::Result<!, u32>` does not permit being left uninitialized

        std::mem::uninitialized::<Option<!>>();
        //~^ ERROR the type `std::option::Option<!>` does not permit being left uninitialized

        std::mem::uninitialized::<OneVariantWith<char>>();
        //~^ ERROR the type `OneVariantWith<char>` does not permit being left uninitialized

        std::mem::uninitialized::<OneVariantWith<!>>();
        //~^ ERROR the type `OneVariantWith<!>` does not permit being left uninitialized

        std::mem::uninitialized::<MaybeUninit<Box<u32>>>();
        std::mem::uninitialized::<usize>();
        std::mem::uninitialized::<f32>();
        std::mem::uninitialized::<*const u8>();
        std::mem::uninitialized::<[u8; 64]>();
        std::mem::uninitialized::<OneVariant>();
        std::mem::uninitialized::<OneVariantWith<u32>>();
    }
}
