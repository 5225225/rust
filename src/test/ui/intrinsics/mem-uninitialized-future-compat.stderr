error: the type `T` is generic, and might not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:13:5
   |
LL |     std::mem::uninitialized::<T>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     this code causes undefined behavior when executed
   |     help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
note: the lint level is defined here
  --> $DIR/mem-uninitialized-future-compat.rs:2:9
   |
LL | #![deny(mem_uninitialized)]
   |         ^^^^^^^^^^^^^^^^^
   = note: for more information, see FIXME: fill this in
   = note: type might not be allowed to be left uninitialized

error: the type `[T; N]` is generic, and might not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:16:5
   |
LL |     std::mem::uninitialized::<[T; N]>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     this code causes undefined behavior when executed
   |     help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
   = note: for more information, see FIXME: fill this in
   = note: type might not be allowed to be left uninitialized

error: the type `[char; N]` is generic, and might not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:19:5
   |
LL |     std::mem::uninitialized::<[char; N]>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     this code causes undefined behavior when executed
   |     help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
   = note: for more information, see FIXME: fill this in
   = note: characters must be a valid Unicode codepoint

error: the type `[UninitStruct; N]` is generic, and might not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:22:5
   |
LL |     std::mem::uninitialized::<[UninitStruct; N]>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     this code causes undefined behavior when executed
   |     help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
   = note: for more information, see FIXME: fill this in
note: characters must be a valid Unicode codepoint (in this struct field)
  --> $DIR/mem-uninitialized-future-compat.rs:9:5
   |
LL |     b: char,
   |     ^^^^^^^

error: the type `&u32` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:31:9
   |
LL |         std::mem::uninitialized::<&'static u32>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
   = note: for more information, see FIXME: fill this in
   = note: references must be non-null

error: the type `std::boxed::Box<u32>` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:34:9
   |
LL |         std::mem::uninitialized::<Box<u32>>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
   = note: for more information, see FIXME: fill this in
   = note: `Box` must be non-null

error: the type `fn()` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:37:9
   |
LL |         std::mem::uninitialized::<fn()>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
   = note: for more information, see FIXME: fill this in
   = note: function pointers must be non-null

error: the type `!` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:40:9
   |
LL |         std::mem::uninitialized::<!>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
   = note: for more information, see FIXME: fill this in
   = note: the `!` type has no valid value

error: the type `*mut dyn std::io::Write` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:43:9
   |
LL |         std::mem::uninitialized::<*mut dyn std::io::Write>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
   = note: for more information, see FIXME: fill this in
   = note: the vtable of a wide raw pointer must be non-null

error: the type `bool` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:46:9
   |
LL |         std::mem::uninitialized::<bool>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
   = note: for more information, see FIXME: fill this in
   = note: booleans must be either `true` or `false`

error: the type `char` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:49:9
   |
LL |         std::mem::uninitialized::<char>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
   = note: for more information, see FIXME: fill this in
   = note: characters must be a valid Unicode codepoint

error: the type `UninitStruct` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:52:9
   |
LL |         std::mem::uninitialized::<UninitStruct>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
   = note: for more information, see FIXME: fill this in
note: characters must be a valid Unicode codepoint (in this struct field)
  --> $DIR/mem-uninitialized-future-compat.rs:9:5
   |
LL |     b: char,
   |     ^^^^^^^

error: the type `[UninitStruct; 16]` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:55:9
   |
LL |         std::mem::uninitialized::<[UninitStruct; 16]>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
   = note: for more information, see FIXME: fill this in
note: characters must be a valid Unicode codepoint (in this struct field)
  --> $DIR/mem-uninitialized-future-compat.rs:9:5
   |
LL |     b: char,
   |     ^^^^^^^

error: the type `(u32, char)` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:58:9
   |
LL |         std::mem::uninitialized::<(u32, char)>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
   = note: for more information, see FIXME: fill this in
   = note: characters must be a valid Unicode codepoint

error: aborting due to 14 previous errors

Future incompatibility report: Future breakage diagnostic:
error: the type `T` is generic, and might not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:13:5
   |
LL |     std::mem::uninitialized::<T>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     this code causes undefined behavior when executed
   |     help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
note: the lint level is defined here
  --> $DIR/mem-uninitialized-future-compat.rs:2:9
   |
LL | #![deny(mem_uninitialized)]
   |         ^^^^^^^^^^^^^^^^^
   = note: for more information, see FIXME: fill this in
   = note: type might not be allowed to be left uninitialized

Future breakage diagnostic:
error: the type `[T; N]` is generic, and might not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:16:5
   |
LL |     std::mem::uninitialized::<[T; N]>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     this code causes undefined behavior when executed
   |     help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
note: the lint level is defined here
  --> $DIR/mem-uninitialized-future-compat.rs:2:9
   |
LL | #![deny(mem_uninitialized)]
   |         ^^^^^^^^^^^^^^^^^
   = note: for more information, see FIXME: fill this in
   = note: type might not be allowed to be left uninitialized

Future breakage diagnostic:
error: the type `[char; N]` is generic, and might not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:19:5
   |
LL |     std::mem::uninitialized::<[char; N]>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     this code causes undefined behavior when executed
   |     help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
note: the lint level is defined here
  --> $DIR/mem-uninitialized-future-compat.rs:2:9
   |
LL | #![deny(mem_uninitialized)]
   |         ^^^^^^^^^^^^^^^^^
   = note: for more information, see FIXME: fill this in
   = note: characters must be a valid Unicode codepoint

Future breakage diagnostic:
error: the type `[UninitStruct; N]` is generic, and might not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:22:5
   |
LL |     std::mem::uninitialized::<[UninitStruct; N]>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     this code causes undefined behavior when executed
   |     help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
note: the lint level is defined here
  --> $DIR/mem-uninitialized-future-compat.rs:2:9
   |
LL | #![deny(mem_uninitialized)]
   |         ^^^^^^^^^^^^^^^^^
   = note: for more information, see FIXME: fill this in
note: characters must be a valid Unicode codepoint (in this struct field)
  --> $DIR/mem-uninitialized-future-compat.rs:9:5
   |
LL |     b: char,
   |     ^^^^^^^

Future breakage diagnostic:
error: the type `&u32` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:31:9
   |
LL |         std::mem::uninitialized::<&'static u32>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
note: the lint level is defined here
  --> $DIR/mem-uninitialized-future-compat.rs:2:9
   |
LL | #![deny(mem_uninitialized)]
   |         ^^^^^^^^^^^^^^^^^
   = note: for more information, see FIXME: fill this in
   = note: references must be non-null

Future breakage diagnostic:
error: the type `std::boxed::Box<u32>` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:34:9
   |
LL |         std::mem::uninitialized::<Box<u32>>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
note: the lint level is defined here
  --> $DIR/mem-uninitialized-future-compat.rs:2:9
   |
LL | #![deny(mem_uninitialized)]
   |         ^^^^^^^^^^^^^^^^^
   = note: for more information, see FIXME: fill this in
   = note: `Box` must be non-null

Future breakage diagnostic:
error: the type `fn()` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:37:9
   |
LL |         std::mem::uninitialized::<fn()>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
note: the lint level is defined here
  --> $DIR/mem-uninitialized-future-compat.rs:2:9
   |
LL | #![deny(mem_uninitialized)]
   |         ^^^^^^^^^^^^^^^^^
   = note: for more information, see FIXME: fill this in
   = note: function pointers must be non-null

Future breakage diagnostic:
error: the type `!` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:40:9
   |
LL |         std::mem::uninitialized::<!>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
note: the lint level is defined here
  --> $DIR/mem-uninitialized-future-compat.rs:2:9
   |
LL | #![deny(mem_uninitialized)]
   |         ^^^^^^^^^^^^^^^^^
   = note: for more information, see FIXME: fill this in
   = note: the `!` type has no valid value

Future breakage diagnostic:
error: the type `*mut dyn std::io::Write` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:43:9
   |
LL |         std::mem::uninitialized::<*mut dyn std::io::Write>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
note: the lint level is defined here
  --> $DIR/mem-uninitialized-future-compat.rs:2:9
   |
LL | #![deny(mem_uninitialized)]
   |         ^^^^^^^^^^^^^^^^^
   = note: for more information, see FIXME: fill this in
   = note: the vtable of a wide raw pointer must be non-null

Future breakage diagnostic:
error: the type `bool` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:46:9
   |
LL |         std::mem::uninitialized::<bool>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
note: the lint level is defined here
  --> $DIR/mem-uninitialized-future-compat.rs:2:9
   |
LL | #![deny(mem_uninitialized)]
   |         ^^^^^^^^^^^^^^^^^
   = note: for more information, see FIXME: fill this in
   = note: booleans must be either `true` or `false`

Future breakage diagnostic:
error: the type `char` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:49:9
   |
LL |         std::mem::uninitialized::<char>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
note: the lint level is defined here
  --> $DIR/mem-uninitialized-future-compat.rs:2:9
   |
LL | #![deny(mem_uninitialized)]
   |         ^^^^^^^^^^^^^^^^^
   = note: for more information, see FIXME: fill this in
   = note: characters must be a valid Unicode codepoint

Future breakage diagnostic:
error: the type `UninitStruct` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:52:9
   |
LL |         std::mem::uninitialized::<UninitStruct>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
note: the lint level is defined here
  --> $DIR/mem-uninitialized-future-compat.rs:2:9
   |
LL | #![deny(mem_uninitialized)]
   |         ^^^^^^^^^^^^^^^^^
   = note: for more information, see FIXME: fill this in
note: characters must be a valid Unicode codepoint (in this struct field)
  --> $DIR/mem-uninitialized-future-compat.rs:9:5
   |
LL |     b: char,
   |     ^^^^^^^

Future breakage diagnostic:
error: the type `[UninitStruct; 16]` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:55:9
   |
LL |         std::mem::uninitialized::<[UninitStruct; 16]>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
note: the lint level is defined here
  --> $DIR/mem-uninitialized-future-compat.rs:2:9
   |
LL | #![deny(mem_uninitialized)]
   |         ^^^^^^^^^^^^^^^^^
   = note: for more information, see FIXME: fill this in
note: characters must be a valid Unicode codepoint (in this struct field)
  --> $DIR/mem-uninitialized-future-compat.rs:9:5
   |
LL |     b: char,
   |     ^^^^^^^

Future breakage diagnostic:
error: the type `(u32, char)` does not permit being left uninitialized
  --> $DIR/mem-uninitialized-future-compat.rs:58:9
   |
LL |         std::mem::uninitialized::<(u32, char)>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         this code causes undefined behavior when executed
   |         help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
   |
note: the lint level is defined here
  --> $DIR/mem-uninitialized-future-compat.rs:2:9
   |
LL | #![deny(mem_uninitialized)]
   |         ^^^^^^^^^^^^^^^^^
   = note: for more information, see FIXME: fill this in
   = note: characters must be a valid Unicode codepoint

